flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/reconciler/deployment/AbstractFlinkResourceReconciler.java

```diff
@@ -186,7 +187,41 @@ private void applyAutoscaler(FlinkResourceContext<CR> ctx) throws Exception {
                 ctx.getResource().getSpec().getJob() != null
                         && ctx.getObserveConfig().getBoolean(AUTOSCALER_ENABLED);
         autoScalerCtx.getConfiguration().set(AUTOSCALER_ENABLED, autoscalerEnabled);
+
         autoscaler.scale(autoScalerCtx);
+        putBackOldParallelismOverridesIfNewOnesAreMerelyAPermutation(ctx);
+    }
+
+    private static <
+                    CR extends AbstractFlinkResource<SPEC, STATUS>,
+                    SPEC extends AbstractFlinkSpec,
+                    STATUS extends CommonStatus<SPEC>>
+            void putBackOldParallelismOverridesIfNewOnesAreMerelyAPermutation(
```

Comment:
Could we move this logic to the `ScalingRealizer`? https://github.com/apache/flink-kubernetes-operator/commit/158cbe29169cbfb7fa7ad676fb0273fd7ef6d25e adds the logic there and this issue comes from that change. I feel these 2 changes logically belong together and it's weird that we break something in once place and fix it in another while it could be simply next to each other.

It will likely also make the whole logic a bit simpler

Reply:
It's not possible to add the change there because we do not have access to the current overrides.

Reply:
Also https://github.com/apache/flink-kubernetes-operator/commit/158cbe29169cbfb7fa7ad676fb0273fd7ef6d25e does not introduce the problem. It fixes it at the cost of restarting once instead of every time we redeploy the operator.